### DISCUSSION TITLE
Java runtime support

### DISCUSSION THREAD

**Discussion by Meldiron** [appwrite/appwrite#2904] [Q&A]
With Appwrite 0.13, a new build and execution model has been applied. This model comes from [Open Runtimes](https://github.com/open-runtimes/open-runtimes), and requires every runtime to have an HTTP server, and dynamically importing user code.

Every language is a little different, and in Java, there are many ways to dynamically import a user's code.

Let's open a discussion what would be the ideal implementation of dynamically importing user's code into Java HTTP server while supporting:

- Custom user code dependencies
- Relative imports in user code

**Comment by Meldiron** [comment]
The first idea that comes to my mind is the `.jar`file. During the build process, we would create a `.jar` file of user's code, and then import it into HTTP server code. Jar files are meant for this, I believe this would be the best solution. If you come across any problems with JAR files, please let me know 🙏

**Reply by yk-sgr** [reply]
Should be absolutely fine.

**Comment by Meldiron** [comment]
A possible problem that comes to my mind is that Java is strictly typed... We will have a type definition on HTTP server for request and response objects (classes), but users would not have these classes in their source code. How can we work around that? Is there some alternative to `any` or `json` in Java? Is it commonly used?

**Reply by maeddin** [reply]
For example, you could use Jackson. With it you can convert any object into JSON and the user could also customize how it is formatted if he wants. Of course, you would have to state in the documentation that Jackson is being used so that the users can understand this.
Jackson is generally very popular for such use cases, for example in Spring.
You could of course also use a Map<String, Object>, but that's not best practice and would be just as possible with Jackson if the users really want it.

**Reply by Meldiron** [reply]
Thanks for the feedback ❤️ I noticed multiple JSON libraries, but none of them was supported in Java by default. This means, we would need to force a dependency on a user, and my goal is to prevent that. The idea is that users will have to define some kind of function in their code:

```
public static void main(TYPE request, TYPE response) {
```

If these types were from a library, the user will need to install this library in their code.

**Reply by yk-sgr** [reply]
You could probably just add an `Serializable` interface or something similar to let the users implement their own serialization strategy.

**Reply by Meldiron** [reply]
Thanks for the suggestion! I tried both `Serializable` and `Object`, but could not figure out how to get data from such a variable. I could re-cast it multiple types, but each user would need to do this in their function and that would get annoying really quick 😅

![CleanShot 2022-03-05 at 10 39 15](https://user-images.githubusercontent.com/19310830/156877712-20df67f5-709c-41ff-9ee4-418364865e50.png)

It is also not clear how we could define an executable method such as `response->json()`.

These are types of request and response we need to achieve:  https://github.com/open-runtimes/open-runtimes/blob/main/runtimes/deno-1.13/server.ts#L21-L36

**Reply by maeddin** [reply]
You can write your own Serializables, of course, but that is absolutely not user-friendly in my eyes.
Since Java is strictly typed, there is no such thing as 'any'. This means that either the user has to use a library of AppWrite that contains the classes, or the 'primitive' objects have to be passed all individually. For most things you can use a Map<String, Object>. The functions like the json method you probably either have to pass in as a function or just let the user return what they want to send, which of course would leave you with nothing to execute after sending. Another option would be to edit a passed object so that you "write" your response in there, but that would be relatively ugly.


The problem is that Java is not designed for something like that. No matter how you do it, it's not pretty. For example, you can put the functions in the map response, because functions are just objects. So you need also in Java only the two objects request and response. But maybe there is an more elegant way, which I don't know or can't think of.

Unfortunately, that's all I can think of.

**Reply by maeddin** [reply]
Without importing an explicit library, you can't get around many casts in Java.
Kotlin, on the other hand, has a dynamic type, but only to be able to interact with JavaScript.

**Comment by Meldiron** [comment]
Yet another question comes to my mind, how do we manage dependencies? I noticed there is Maven, and there is also Gradle. Are there any other dependency managers? Would it be possible to support both?

**Reply by yk-sgr** [reply]
Gradle and Maven are the only relevant ones.

**Reply by Meldiron** [reply]
Thanks for confirmation 👍 There could be a simple if/else in bash during the build process, and run appropriate install commands for each dependency manager.
